### PR TITLE
Add Haiku OS support

### PR DESCRIPTION
**Pull request by avanspector**
core:sync works although has some FIXMEs
core:fmt works
core:sys/haiku is added

generally in Odin Haiku is considered a Unix-like OS, even though technically it is not true

(core:os additions are quite remarkable in their raw quality, albeit being functional enough. As sys/haiku grows, core:os_haiku improves too though)

### CODE REVIEW

**Review by Skytrias — COMMENTED** [review]
small files looking good, does the futex sim have tests or are they implemented the same as something we got already

**Review by Skytrias — COMMENTED** [review]
rest looks fine, good work 👍

File: core/sync/futex_haiku.odin [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+	waiter.next.prev = waiter.prev
+
+ 	unix.pthread_sigmask(haiku.SIG_SETMASK, &old_mask, nil)
+
+ 	// FIXME: Add error handling!
+ 	return true
+}
+
+_futex_wait_with_timeout :: proc "contextless" (f: ^Futex, expect: u32, duration: time.Duration) -> bool {
+	// FIXME: Add timeout!
```
Comment (gingerBill) [inline core/sync/futex_haiku.odin:97]: Make sure that you do! This is necessary for the sync primitives

File: src/path.cpp [hunk 2]
```diff
@@ -1,461 +1,461 @@
-/*
-	Path handling utilities.
-*/
-#if !defined(GB_SYSTEM_WINDOWS)
-#include <unistd.h>
-#endif
```
Comment (gingerBill) [inline src/path.cpp:6]: What happened with this file? What's up with the diff?

File: src/threading.cpp [hunk 3]
```diff
@@ -831,8 +833,119 @@ gb_internal void futex_wait(Futex *f, Footex val) {
...
+	Futex_Wait_Node *prev, *next;	
+};
+ 
+struct Futex_Wait_Queue {
+	std::atomic_flag spinlock;
+	Futex_Wait_Node list;
+ 
+	void lock() {
+		while (spinlock.test_and_set(std::memory_order_acquire)) {
+			; // spin...
```
Comment (gingerBill) [inline src/threading.cpp:856]: At least put a cpu_relax in there of some sort.

File: core/sync/futex_haiku.odin [hunk 4]
```diff
@@ -0,0 +1,127 @@
+//+private
+package sync
+
+import "core:c"
+import "core:c/libc"
```
Comment (gingerBill) [inline core/sync/futex_haiku.odin:5]: Do not use libc here the atomics. package sync has all of the atomics stuff in it already.

File: core/sync/futex_haiku.odin [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+@(private="file")
+Wait_Queue :: struct {
+	lock: libc.atomic_flag,
+	list: Wait_Node,
+}
+@(private="file")
+waitq_lock :: proc "contextless" (waitq: ^Wait_Queue) {
+	// FIXME: Get rid of context here.
+	context = runtime.default_context()
+	for libc.atomic_flag_test_and_set_explicit(&waitq.lock, .acquire) {
```
Comment (gingerBill) [inline core/sync/futex_haiku.odin:26]: Use the atomics built into the language.

File: core/sync/futex_haiku.odin [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+			prev = &g_waitq.list,
+			next = &g_waitq.list,
+		},
+	}
+}
+
+@(private="file")
+get_waitq :: #force_inline proc "contextless" (f: ^Futex) -> ^Wait_Queue {
+	_ = f
+	return &g_waitq
```
Comment (gingerBill) [inline core/sync/futex_haiku.odin:55]: If this is private to the file, why even bother with a dummy parameter?

File: core/sync/futex_haiku.odin [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+_futex_broadcast :: proc "contextless" (f: ^Futex) {
+	waitq := get_waitq(f)
+	waitq_lock(waitq)
+	defer waitq_unlock(waitq)
+
+	head := &waitq.list
+	for waiter := head.next; waiter != head; waiter = waiter.next {
+		if waiter.futex == f {
+			unix.pthread_kill(waiter.thread, haiku.SIGCONT)
```
Comment (gingerBill) [inline core/sync/futex_haiku.odin:124]: Is there not a distinction between signal and broadcast on Haiku?

File: src/threading.cpp (hunk 3 above)
Comment (avanspector) [reply src/threading.cpp:856]: is there an alternative for this in the compiler codebase?

File: src/path.cpp (hunk 2 above)
Comment (avanspector) [reply src/path.cpp:6]: only one `#if defined(GB_SYSTEM_HAIKU)` for `dirent.h` was added, everything else is CRLF conversion

**Comment by avanspector** [comment]
> small files looking good, does the futex sim have tests or are they implemented the same as something we got already

futex for Haiku was simulated using libc, implementation taken from here [](https://tavianator.com/2023/futex.html)
im a novice in the whole futexes/multithreading situation, all test was basically running on the OS, negating chances of deadlocking as much as possible (i'll try reaching out to the haiku devs to make the implementation as reliable as possible since this simulation of futexes works fine on Linux)

File: src/threading.cpp (hunk 3 above)
Comment (gingerBill) [reply src/threading.cpp:856]: `intrinsics.cpu_relax()`

**Pull request merged**